File: 1547.java
class Solution {
    public int minCost(int n, int[] cuts) {
        Arrays.sort(cuts);
        int m = cuts.length;
        int[][] dp = new int[m+1][m+1];
        for (int i = 0; i < m; i++)
            Arrays.fill(dp[i], Integer.MAX_VALUE);
        for (int i = 0; i < m; i++)
            dp[i][i] = 0;
        for(int i = m - 1; i >= 0; i--) {
            for(int j = i + 1; j < m + 1; j++) {
                for(int k = i; k < j; k++) {
                    int cost = (j == m ? n : cuts[j]) - (i == 0 ? 0 : cuts[i-1]);
                    dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k+1][j] + cost);
                }
            }
        }
        return dp[0][m];
    }
}

File: 1547.py
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts.extend([0, n])
        cuts.sort()
        @lru_cache(None)
        def helper(i, j):
            if i + 1 == j:
                return 0
            res = 0x3f3f3f3f
            for k in range(i + 1, j):
                res = min(res, helper(i, k) + helper(k, j) + cuts[j] - cuts[i])
            return res
        return helper(0, len(cuts) - 1)